File: 0984-most-stones-removed-with-same-row-or-column/0984-most-stones-removed-with-same-row-or-column.go
type union struct {
    rank   []int
    parent []int
}

func (u *union) find(x int) int {
    if u.parent[x] != x {
        u.parent[x] = u.find(u.parent[x])
    }
    return u.parent[x]
}

func (u *union) unionSets(x, y int) {
    rootX := u.find(x)
    rootY := u.find(y)

    if rootX != rootY {
        if u.rank[rootX] > u.rank[rootY] {
            u.parent[rootY] = rootX
            u.rank[rootX] += u.rank[rootY]
        } else {
            u.parent[rootX] = rootY
            if u.rank[rootX] == u.rank[rootY] {
                u.rank[rootY]++
            }
        }
    }
}

func removeStones(stones [][]int) int {
    n := len(stones)
    
    maxRow, maxCol := 0, 0
    for _, stone := range stones {
        if stone[0] > maxRow {
            maxRow = stone[0]
        }
        if stone[1] > maxCol {
            maxCol = stone[1]
        }
    }
    
    u := union{
        rank:   make([]int, maxRow+maxCol+2),
        parent: make([]int, maxRow+maxCol+2),
    }

    for i := 0; i <= maxRow+maxCol+1; i++ {
        u.parent[i] = i
    }

    for _, stone := range stones {
        row := stone[0]
        col := stone[1] + maxRow + 1
        u.unionSets(row, col)
    }

    uniqueParents := make(map[int]struct{})
    for _, stone := range stones {
        root := u.find(stone[0])
        uniqueParents[root] = struct{}{}
    }

    return n - len(uniqueParents)
}

File: 0984-most-stones-removed-with-same-row-or-column/0984-most-stones-removed-with-same-row-or-column.cpp
class Union {
public:
    vector<int> rank, parent;
    Union(int n) {
        rank.resize(n+1, 1);
        parent.resize(n+1);
        for(int i = 0; i <= n; i++) {
            parent[i] = i;
        }
    }

    int find(int x) {
        if(x == parent[x]) return x;
        return parent[x] = find(parent[x]);
    }

    void unionSet(int u, int v) {
        int uRoot = find(u);
        int vRoot = find(v);

        if(uRoot == vRoot) return;
        if(rank[uRoot] < rank[vRoot]) {
            parent[uRoot] = vRoot;
            rank[vRoot] += rank[uRoot];
        } else {
            parent[vRoot] = uRoot;
            rank[uRoot] += rank[vRoot];
        }
    }
};

class Solution {
public:
    int removeStones(vector<vector<int>>& stones) {
        int n = stones.size();
        int rMax = 0;
        int cMax = 0;

        for(const auto &stone : stones) {
            rMax = max(rMax, stone[0]);
            cMax = max(cMax, stone[1]);
        }

        Union u(rMax+cMax+1);
        unordered_map<int, int> mp;

        for(auto &stone : stones) {
            int row = stone[0];
            int col = stone[1] + rMax + 1;

            u.unionSet(row, col);
            mp[row] = 1;
            mp[col] = 1;
        }

        int cnt = 0;
        for(const auto &element : mp) {
            if(u.find(element.first) == element.first) cnt++;
        }

        return n-cnt;
    }
};